ext/drpttracer/drpttracer.h

```diff
@@ -0,0 +1,211 @@
+ */
+bool
+drpttracer_init(void);
+
+DR_EXPORT
+/**
+ * Cleans up the drpttracer extension.
+ */
+void
+drpttracer_exit(void);
+
+/***************************************************************************
+ * PT TRACING APIS
+ */
+
+/**
+ * Success code for each drpttracer operation.
+ */
+typedef enum {
+    /** Operation succeeded. */
+    DRPTTRACER_SUCCESS,
+    /** Operation failed. */
+    DRPTTRACER_ERROR,
+    /** Operation failed: invalid tracing handle. */
+    DRPTTRACER_ERROR_INVALID_TRACING_HANDLE,
+    /** Operation failed: invalid tracing mode. */
+    DRPTTRACER_ERROR_INVALID_TRACING_MODE,
+    /** Operation failed: failed to open perf event. */
+    DRPTTRACER_ERROR_FAILED_TO_OPEN_PERF_EVENT,
+    /** Operation failed: failed to create tracer handle. */
+    DRPTTRACER_ERROR_FAILED_TO_CREATE_PTTRACER_HANDLE,
+    /** Operation failed: failed to start tracing. */
+    DRPTTRACER_ERROR_FAILED_TO_START_TRACING,
+    /** Operation failed: failed to stop tracing. */
+    DRPTTRACER_ERROR_FAILED_TO_STOP_TRACING
+} drpttracer_status_t;
+
+/**
+ * The tracing modes for drpttracer_start_tracing().
+ */
+typedef enum {
+    /* drpttracer only traces in user space. */
+    DRPTTRACER_TRACING_ONLY_USER,
+    /* drpttracer only traces in kernel. */
+    DRPTTRACER_TRACING_ONLY_KERNEL,
+    /* drpttracer traces both user and kernel. */
+    DRPTTRACER_TRACING_USER_AND_KERNEL
+} drpttracer_tracing_mode_t;
+
+DR_EXPORT
+/**
+ * Starts PT tracing. Must be called after drpttracer_init() and before drpttracer_exit().
+ *
+ * \param[in] drcontext  The context of DynamoRIO.
+ * \param[in] mode  The tracing mode.
+ * \param[out] tracer_handle  The tracer handle.
+ *
+ * \note Each tracing corresponds to a trace_handle. When calling
+ * drpttracer_start_tracing(), the caller will get a tracer_handle. And the caller can
+ * stop tracing by passing it to drpttracer_stop_tracing().
+ *
+ * \return the status code.
+ */
+drpttracer_status_t
+drpttracer_start_tracing(IN void *drcontext, IN drpttracer_tracing_mode_t mode,
+                         OUT void **tracer_handle);
+
+DR_EXPORT
+/**
+ * Stops PT tracing.
+ *
+ * \param[in] drcontext  The context of DynamoRIO.
+ * \param[in] tracer_handle  The tracer handle.
+ * \param[out] meta_data  The metadata of the PT trace. If NULL, the metadata will not be
+ * return.
+ * \param[out] pt_data  The buffer of the PT trace. If NULL, the PT trace will not be
+ * return.
+ * \param[out] sideband_data  The buffer of the PT sideband data. If NULL, the PT sideband
+ * data will not be return.
+ *
+ * \return the status code.
+ */
+drpttracer_status_t
+drpttracer_end_tracing(IN void *drcontext, IN void *tracer_handle,
+                       OUT pt_metadata_t *meta_data, OUT drpttracer_buf_t *pt_data,
+                       OUT drpttracer_buf_t *sideband_data);
+
+DR_EXPORT
+/**
+ * Creates a new drpttracer_buf_t.
+ *
+ * \param[in] drcontext  The context of DynamoRIO.
+ *
+ * \return the created buffer.
```

Comment:
I do not understand this: the docs do not say where/when to call this; it does not take in a size.  How does the user specify a size?  Is it pttracer who controls the size?  If pttracer controls it, why does this create_buffer function exist?

Reply:
Remove this function. Use drpttracer_output_t to store all output data. And the output data struct will be allocated and initialized by drpttracer.